risc0/zkvm/src/exec/executor.rs

```diff
@@ -197,7 +198,32 @@ impl<'a> Executor<'a> {
 
     /// Run the executor until [ExitCode::Paused] or [ExitCode::Halted] is
     /// reached, producing a [Session] as a result.
-    pub fn run_with_callback<F>(&mut self, mut callback: F) -> Result<Session>
+    pub fn run_with_callback<F>(&mut self, callback: F) -> Result<Session>
+    where
+        F: FnMut(Segment) -> Result<Box<dyn SegmentRef>>,
+    {
+        let mut guest_session = self.run_guest_only_with_callback(callback)?;
+        match guest_session.exit_code {
+            ExitCode::Fault => {
+                let fault_checker_session = self.run_fault_checker()?;
+                for segment in fault_checker_session.segments {
+                    guest_session.segments.push(segment);
+                }
+                guest_session.journal = fault_checker_session.journal;
+                Ok(guest_session)
+            }
+            _ => Ok(guest_session),
+        }
+    }
+
+    /// Run the executor with the default callback.
+    pub fn run_guest_only(&mut self) -> Result<Session> {
```

Comment:
My assumption is that the default will be for the fault checker to run if we detect a fault. It's not clear why a user would want to toggle this behavior, to me it's part of the system.

Reply:
So my thinking here that that we definitely want to detect and signal a fault to the application developer. Using the `Fault` exit code seems like a good way to do that, which in my mind is a system split induced by a detected fault condition rather than running out of cycles in the segment.

The distinction I am drawing is that detecting a fault does not necessarily mean running the fault checker. Detection already happens in the executor, and so without running the fault checker this condition can be raised to the developer. My assumption right now is that upon fault, most applications will want to abort proving. Running the fault checker guest program is only needed if the goal is then to prove that a fault occurred. This is why my thought is that upon the executor detecting fault, it would be to rewind and mark the exit status as `Fault`, then provide the developer a way to proving the fault, and so running the fault checker guest, if they have that as part of their error handling path.